### ISSUE TITLE
Unexpected behavior for ternary operators with implicit casting of null

### ISSUE THREAD

**Issue by odebastiani** [dotnet/roslyn#81458]
**Brief description:**

The `test` variable is in the following code not set to `null`, which is unexpected:
```
ReadOnlyMemory<byte>? test = (true ? null : ReadOnlyMemory<byte>.Empty);
```

I therefore suggest a warning for this case to avoid this unexpected behavior. The specific reason is described below.

**Languages applicable:**

C# and probably VB.

**Code example that the analyzer should report:**

I have the following code and expected that `testNull` would be `null` and `testNotNull` would not be `null`:
```
bool testBool = true;
ReadOnlyMemory<byte> defaultValue = new ReadOnlyMemory<byte>([1, 2]);

ReadOnlyMemory<byte>? testNull = (testBool ? null : defaultValue);
ReadOnlyMemory<byte>? testNotNull = (testBool ? defaultValue : null);
```

Surprisingly, this is not the case. Both variables are not `null` and the compiler recognizes that and raises warnings if I check for `null`.
```
bool testNullIsNull = (testNull == null); // CA15508: 'test1 == null' is always 'false'. Remove or refactor the condition(s) to avoid dead code.
bool testNotNullIsNull = (testNotNull == null); // CA15508: 'test2 == null' is always 'false'. Remove or refactor the condition(s) to avoid dead code.
```

The reason for this behavior are a combination of `struct` vs `class` and the `implicit operator`s in `ReadOnlyMemory<T>` which are unfortunately applied to the `null` literal. The same code but with a `class` instead of a `struct` would work as expected.

Here is a simplified example:
```
using System;

namespace Test
{
    public struct TestStruct
    {
        public static implicit operator TestStruct(AccessViolationException? obj) // can be any class or nullable struct
        {
            return new TestStruct();
        }
    }

    public class TestClass
    {
        public static implicit operator TestClass(AccessViolationException? obj) // can be any class or struct
        {
            return new TestClass();
        }
    }

    public static class Program
    {
        public static void Main()
        {
            UseTestStruct(true);
            UseTestClass(true);
        }

        private static void UseTestStruct(bool testBool)
        {
            TestStruct defaultValue = new TestStruct();

            TestStruct? testNull = (testBool ? null : defaultValue);
            TestStruct? testNotNull = (testBool ? defaultValue : null);

            // UNEXPECTED: testNull expected to be null but is default(TestStruct)
            bool testNullIsNull = (testNull == null); // CA1508: 'testNull == null' is always 'false'. Remove or refactor the condition(s) to avoid dead code.

            // UNEXPECTED: testNotNull expected to be not null (correct) but the warning
            //             is wrong because is could be null if testBool was false.
            bool testNotNullIsNull = (testNotNull == null); // CA1508: 'testNotNull == null' is always 'false'. Remove or refactor the condition(s) to avoid dead code.

            Console.WriteLine(testNullIsNull + ", " + testNotNullIsNull);
        }

        private static void UseTestClass(bool testBool)
        {
            TestClass defaultValue = new TestClass();

            TestClass? testNull = (testBool ? null : defaultValue);
            TestClass? testNotNull = (testBool ? defaultValue : null);

            bool testNullIsNull = (testNull == null); // true and no warning
            bool testNotNullIsNull = (testNotNull == null); // false and no warning

            Console.WriteLine(testNullIsNull + ", " + testNotNullIsNull);
        }
    }
}
```

If `TestStruct` had an implicit operator only for a non-nullable `struct` like `int`, the code would work as expected (`testNull` would be `null`).

To fix the code, `(TestStruct?)null` instead of just `null` can be used in the ternary operator.

It is unexpected that implicit operators are used for `null` literals in ternary operators. While this behavior is probably _correct_ (or can't be changed without breaking existing code) it would be very useful to see a warning to add a little fix (explicit casting).

I therefore suggest a warning for the following case:
- for ternary operators
- with a nullable `struct` as target type
- with `null` literal as one case
- target type has `implicit operator` with nullable type

**Additional information:**

--

**Documentation requirements:**

--

**Comment by huoyaoyuan** [comment]
The expectation is incorrect. Ternary expression doesn't apply target type to the two arms. Instead, the best common type is decided for the two arms. The `null` is first converted to the type of LHS of colon `ReadOnlyMemory<byte>`, then the whole ternary expression is casted to `ReadOnlyMemory<byte>`.

To demonstrate more:

```csharp
int? a = false ? 0 : null; // Doesn't work in C# 8 - the best common type of value type and null literal was added in C# 9

int? b = false ? 0 : default; // The `default` literal resolves to default(int), not default(int?)
```

Related to https://github.com/dotnet/csharplang/discussions/8288.
